graphbolt/src/cuda/unique_and_compact_impl.cu

```diff
@@ -0,0 +1,161 @@
+        auto only_src_size =
+            thrust::remove_copy_if(
+                exec_policy, src_ids_ptr, src_ids_ptr + src_ids.size(0),
+                is_dst.get(), only_src.get(), thrust::identity<bool>{}) -
+            only_src.get();
+        auto sorted_only_src =
+            allocator.AllocateStorage<scalar_t>(only_src_size);
+        {
+          size_t workspace_size;
+          CUDA_CALL(cub::DeviceRadixSort::SortKeys(
+              nullptr, workspace_size, only_src.get(), sorted_only_src.get(),
+              only_src_size, 0, num_bits, stream));
+          auto temp = allocator.AllocateStorage<char>(workspace_size);
+          CUDA_CALL(cub::DeviceRadixSort::SortKeys(
+              temp.get(), workspace_size, only_src.get(), sorted_only_src.get(),
+              only_src_size, 0, num_bits, stream));
+        }
+        auto unique_only_src = torch::empty(only_src_size, src_ids.options());
+        auto unique_only_src_ptr = unique_only_src.data_ptr<scalar_t>();
+        auto unique_only_src_cnt = allocator.AllocateStorage<scalar_t>(1);
+        {
+          size_t workspace_size;
+          CUDA_CALL(cub::DeviceRunLengthEncode::Encode(
+              nullptr, workspace_size, sorted_only_src.get(),
+              unique_only_src_ptr, cub::DiscardOutputIterator{},
+              unique_only_src_cnt.get(), only_src_size, stream));
+          auto temp = allocator.AllocateStorage<char>(workspace_size);
+          CUDA_CALL(cub::DeviceRunLengthEncode::Encode(
+              temp.get(), workspace_size, sorted_only_src.get(),
+              unique_only_src_ptr, cub::DiscardOutputIterator{},
+              unique_only_src_cnt.get(), only_src_size, stream));
+        }
+        scalar_t unique_only_src_size;
+        CUDA_CALL(cudaMemcpyAsync(
+            &unique_only_src_size, unique_only_src_cnt.get(),
+            sizeof(unique_only_src_size), cudaMemcpyDeviceToHost, stream));
+        stream.synchronize();
+        unique_only_src = unique_only_src.slice(0, 0, unique_only_src_size);
+        auto real_order = torch::cat({unique_dst_ids, unique_only_src});
+        auto [sorted_order, new_ids] = Sort(real_order, num_bits);
+        auto sorted_order_ptr = sorted_order.data_ptr<scalar_t>();
+        auto new_ids_ptr = new_ids.data_ptr<int64_t>();
+        auto new_src_ids_loc =
+            allocator.AllocateStorage<scalar_t>(src_ids.size(0));
+        auto new_dst_ids_loc =
+            allocator.AllocateStorage<scalar_t>(dst_ids.size(0));
+        thrust::lower_bound(
+            exec_policy, sorted_order_ptr,
+            sorted_order_ptr + sorted_order.size(0), src_ids_ptr,
+            src_ids_ptr + src_ids.size(0), new_src_ids_loc.get());
+        thrust::lower_bound(
+            exec_policy, sorted_order_ptr,
+            sorted_order_ptr + sorted_order.size(0), dst_ids_ptr,
+            dst_ids_ptr + dst_ids.size(0), new_dst_ids_loc.get());
+        {  // Checks if unique_dst_ids includes all dst_ids
```

Comment:
Why do we need to add a check?

Reply:
Because the tests expect to throw an exception when unique_dst_ids do not include all dst_ids.

Reply:
This check adds overhead, but we are doing it to pass the test and make the code more robust and match the behaviour of CPU algorithm.

Reply:
CPU algorithm here: https://github.com/dmlc/dgl/blob/26e740a3530f0b4a95b48e237c24cefe62bc8f1b/graphbolt/src/unique_and_compact.cc#L53